rdkit/Chem/AllChem.py

```diff
@@ -464,47 +465,164 @@ def AssignBondOrdersFromTemplate(refmol, mol):
-        b.SetIsAromatic(False)
+      b.SetBondType(BondType.SINGLE)
+      b.SetIsAromatic(False)
     # set the bonds of mol to SINGLE
     for b in refmol2.GetBonds():
       b.SetBondType(BondType.SINGLE)
       b.SetIsAromatic(False)
     # set atom charges to zero;
     for a in refmol2.GetAtoms():
       a.SetFormalCharge(0)
+      a.SetIsAromatic(False)
     for a in mol2.GetAtoms():
       a.SetFormalCharge(0)
-      
-    matching = mol2.GetSubstructMatches(refmol2, uniquify=False)
+      a.SetIsAromatic(False)
+
+    matching = mol2.GetSubstructMatch(refmol2)
+    # an inveerse match
+    if not matching:
+      inverse_matching = refmol2.GetSubstructMatch(mol2)
+      if inverse_matching and mol2.GetNumAtoms() == refmol2.GetNumAtoms():
+        matching = tuple(i[0] for i in sorted(enumerate(inverse_matching),key=lambda x: x[1]))
+
+    # add missing and remove excesive bonds
+    if len(GetMolFrags(mol2, asMols=False, sanitizeFrags=False)) > 1 or mol2.GetNumAtoms() != refmol2.GetNumAtoms():
+      fragments = zip(GetMolFrags(mol2, asMols=False, sanitizeFrags=False), GetMolFrags(mol2, asMols=True, sanitizeFrags=False))
+      fragments = sorted(fragments, key=lambda x: len(x[0]), reverse=True) # sort by decreasing size
+
+      available_atoms = set(range(refmol2.GetNumAtoms()))
+      frag_based_matching = []
+      for f, f_mol in fragments:
+        frag_matches = refmol2.GetSubstructMatches(f_mol)
+        for m in frag_matches:
+          if set(m).issubset(available_atoms):
+            available_atoms -= set(m)
+            frag_based_matching.extend(zip(f, m))
+            break
+        else:
+          pass
+      if frag_based_matching and len(frag_based_matching) == refmol2.GetNumAtoms():
+        mol_matching = tuple(i[0] for i in sorted(frag_based_matching,key=lambda x: x[1]))
+        ref_matching = tuple(i[1] for i in sorted(frag_based_matching,key=lambda x: x[0]))
+
+        mol2 = RWMol(mol2)
+
+        missing_bonds=[]
+        for b in refmol2.GetBonds():
+          atom1 = mol_matching[b.GetBeginAtomIdx()]
+          atom2 = mol_matching[b.GetEndAtomIdx()]
+          b2 = mol2.GetBondBetweenAtoms(atom1, atom2)
+          if not b2:
+            missing_bonds.append((atom1, atom2))
+        for a1, a2 in missing_bonds:
+          mol2.AddBond(a1, a2)
+
+        extra_bonds=[]
+        for b in mol2.GetBonds():
+          atom1 = ref_matching[b.GetBeginAtomIdx()]
+          atom2 = ref_matching[b.GetEndAtomIdx()]
+          b2 = refmol2.GetBondBetweenAtoms(atom1, atom2)
+          if not b2:
+            extra_bonds.append((b.GetBeginAtomIdx(), b.GetEndAtomIdx()))
+
+        for a1, a2 in extra_bonds:
+          mol2.RemoveBond(a1, a2)
+
+        mol2 = mol2.GetMol()
+
+        # use current mapping if it's complete or check again
+        if len(mol_matching) != refmol2.GetNumAtoms():
+          matching = mol2.GetSubstructMatch(refmol2)
+
+    if not matching: # MCS - last resort
```

Comment:
Using the MCS as part of this process is an interesting idea, but I think it should be an option (and probably default to False, but given some real examples my mind here can be changed).